merlion/models/layers.py

```diff
@@ -0,0 +1,253 @@
+    for `DetectorBase` or `ForecastingDetectorBase`. The abstract methods (``forecast`` and ``get_anomaly_score``)
+    are overridden to call the underlying model.
+
+    If the base model is a forecaster, the top-level config ``model.config`` does not duplicate attributes of the
+    underlying forecaster config (e.g. ``max_forecast_steps`` or ``target_seq_index``). Instead,
+    ``model.config.max_forecast_steps`` will resolve to ``model.config.base_model.max_forecast_steps``.
+    As a result, you will only need to specify this parameter once. The same holds true for `DetectorConfig` attributes
+    (e.g. ``threshold`` or ``calibrator``) when the base model is an anomaly detector.
+
+    .. note::
+
+        For the time being, every layer of the model is allowed to have its own ``transform``.
+    """
+
+    config_class = LayeredModelConfig
+    require_even_sampling = False
+    require_univariate = False
+
+    def __new__(cls, config: LayeredModelConfig = None, model: ModelBase = None, **kwargs):
+        # Dynamically inherit from the appropriate kind of base model
+        original_cls = cls
+        config = cls._resolve_args(config=config, model=model, **kwargs)
+        if isinstance(config.model, ForecastingDetectorBase):
+            cls = cls.__class__(cls.__name__, (cls, LayeredForecastingDetector), {})
+            setattr(cls, "_original_cls", original_cls)
+        elif isinstance(config.model, ForecasterBase):
+            cls = cls.__class__(cls.__name__, (cls, LayeredForecaster), {})
+            setattr(cls, "_original_cls", original_cls)
+        elif isinstance(config.model, DetectorBase):
+            cls = cls.__class__(cls.__name__, (cls, LayeredDetector), {})
+            setattr(cls, "_original_cls", original_cls)
+        return super().__new__(cls)
+
+    def __init__(self, config: LayeredModelConfig = None, model: ModelBase = None, **kwargs):
+        super().__init__(config=self._resolve_args(config=config, model=model, **kwargs))
+
+    @classmethod
+    def _resolve_args(cls, config: LayeredModelConfig, model: ModelBase, **kwargs):
+        msg = f"Expected exactly one of `config.model` or `model` when creating LayeredModel {cls.__name__}."
+        if config is None and model is None:
+            raise RuntimeError(f"{msg} Received neither.")
+        elif config is not None and model is not None:
+            if config.model is None:
+                config.model = model
+            else:
+                raise RuntimeError(f"{msg} Received both.")
+        elif config is None:
+            config = cls.config_class(model=model, **kwargs)
+        return config
+
+    @property
+    def model(self):
+        return self.config.model
+
+    @model.setter
+    def model(self, model):
+        self.config.model = model
+
+    @property
+    def base_model(self):
+        return self.config.base_model
+
+    def reset(self):
+        self.model.reset()
+        self.__init__(config=self.config)
```

Comment:
Worth pointing out that this isn't going to reset the underlying model (the model is recreated with the config, and as the model is part of the config, the underlying models won't be changed).

Reply:
`self.model` is a reference to `self.config.model`, not a duplicated code object.  So the underlying model actually is reset before the layered model.